### PR TITLE
Manage session id using random int for gradio local mode

### PR DESCRIPTION
**Pull request by aisensiy**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily receiving feedbacks. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

There is a related issue https://github.com/InternLM/lmdeploy/issues/503, right now the solution using host ip will make the session break randomly if the gradio service is behind a stateless load balancer.

## Modification

Add a session_id `State` in gradio, the `State will` bind with current page which will not be influence by host ip from `gradio.Request`.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by AllentDan** [comment]
> This PR have great conflict with #544 ...

Yes, could you also make a support to `api_server` mode?

**Comment by lvhan028** [comment]
@aisensiy community is first :)
@AllentDan Let's leave the conflicts to ourselves 😺

**Comment by irexyc** [comment]
state_session_id seems equal for every session. I think use a global variable is a better idea.

```python
# global session id
global_session_id = [0]

async def chat_stream_local(
    instruction: str,
    state_chatbot: Sequence,
    cancel_btn: gr.Button,
    reset_btn: gr.Button,
    session_id: int,
):
    if session_id == -1:
        session_id = global_session_id[-1] + 1
        global_session_id[-1] += 1

with gr.Blocks(css=CSS, theme=THEME) as demo:
    # state_session_id = gr.State(random.randint(0, 100000))
    state_session_id = gr.State(-1)


```

**Comment by AllentDan** [comment]
其实 app.py 里也有类似 global 的全局变量管理，你如果不适应，也可以放一个 `session_id` 变量到 `InterFace` 统一管理
https://github.com/InternLM/lmdeploy/blob/186bfd2eb9ee44db129d1dac70dcabe01461227b/lmdeploy/serve/gradio/app.py#L149

**Comment by lvhan028** [comment]
Hi, @aisensiy, sorry for keeping you waiting for such a long time.
We finally finished the review of PR #544 
Let's work on this PR. 
Please resolve the conflicts before we moving on

**Comment by AllentDan** [comment]
@aisensiy Hi, we finally merged #544 into main. Could you help resolve conflicts?

**Comment by lvhan028** [comment]
@aisensiy Could you fixed it this week? We plan to release a new version next Monday. And we hope to merge this PR before the release. 
If you are not available, @AllentDan can take it over. 
Do worry about your credit. Your commits history will not be overwritten

**Comment by AllentDan** [comment]
I made some changes to this PR. Please review them.

**Comment by irexyc** [comment]
api_server_backend.py、triton_server_backend.py 这俩文件里面docstrings可以更新一下，比如：
`request (gr.Request): the request from a user` -> `session_id (int): the session id`

File: lmdeploy/serve/gradio/turbomind_coupled.py [hunk 1]
```diff
@@ -166,24 +147,33 @@ def run_local(model_path: str,
...
+            [state_chatbot, cancel_btn, reset_btn],
+            cancels=[send_event])
+
+        reset_btn.click(reset_local_func,
+                        [instruction_txtbox, state_chatbot, state_session_id],
                         [state_chatbot, chatbot, instruction_txtbox],
                         cancels=[send_event])
 
+        def init():
+            InterFace.global_session_id += 1
```
Comment (lvhan028) [inline lmdeploy/serve/gradio/turbomind_coupled.py:171]: When two requests arrive simultaneously, how can we ensure that the access to global_session_id is mutually exclusive?

Comment (AllentDan) [reply lmdeploy/serve/gradio/turbomind_coupled.py:171]: As far as I know, gradio may process requests asynchronously. But anyway, I will add a lock to ensure it is safe.

Comment (irexyc) [reply lmdeploy/serve/gradio/turbomind_coupled.py]: https://www.gradio.app/guides/interface-state#global-state

官方给的全局变量的例子，看用法是线程安全的。

**Review by irexyc — APPROVED** [review]

**Review by lvhan028 — APPROVED** [review]

**Pull request merged**